synthesizer/program/src/resources/credits.aleo

```diff
@@ -106,573 +149,515 @@ record credits:
-    cast 0u64 true into r9 as committee_state;
+    cast true r3 into r5 as committee_state;
     // Retrieve the committee state of the specified validator.
-    get.or_use committee[r0] r9 into r10;
+    get.or_use committee[r0] r5 into r6;
+
+    // Fail if attempting to change commission. Commission can only be set when validator bonds the first time.
+    assert.eq r3 r6.commission;
 
-    // Increment the stake for the specified validator.
-    add r10.microcredits r3 into r11;
     // Construct the updated committee state.
-    cast r11 r10.is_open into r12 as committee_state;
+    cast r6.is_open r6.commission into r7 as committee_state;
 
     /* Bonded */
 
     // Construct the initial bond state.
-    cast r0 0u64 into r13 as bond_state;
-    // Get the bond state for the caller, or default to the initial bond state.
-    get.or_use bonded[r0] r13 into r14;
+    cast r0 0u64 into r8 as bond_state;
+    // Get the bond state for the signer, or default to the initial bond state.
+    get.or_use bonded[r0] r8 into r9;
     // Enforce the validator matches in the bond state.
-    assert.eq r14.validator r0;
+    assert.eq r9.validator r0;
 
     // Increment the microcredits in the bond state.
-    add r14.microcredits r3 into r15;
+    add r9.microcredits r2 into r10;
+
+    // Include total pre-delegated amount when deciding if validator has enough credits to join committee.
+    // First get the current delegated amount, or start with 0 if none delegated.
+    get.or_use delegated[r0] 0u64 into r11;
+    // Self-bond amount + existing delegated amount = total bond.
+    add r2 r11 into r12;
     // Determine if the amount is at least 10 million credits.
-    gte r15 10_000_000_000_000u64 into r16;
+    gte r12 10_000_000_000_000u64 into r13;
     // Enforce the amount is at least 10 million credits.
-    assert.eq r16 true;
+    assert.eq r13 true;
 
     // Construct the updated bond state.
-    cast r0 r15 into r17 as bond_state;
+    cast r0 r10 into r14 as bond_state;
 
     /* Account */
 
-    // Get the balance of the caller.
+    // Get the balance of the signer.
     // If the account does not exist, this finalize scope will fail.
-    get account[r0] into r18;
-    // Decrement the balance of the caller.
-    sub r18 r3 into r19;
+    get account[r0] into r15;
+    // Decrement the balance of the signer.
+    sub r15 r2 into r16;
 
     /* Writes */
 
-    // Update the committee state of the specified validator.
-    set r12 into committee[r0];
-    // Update the bond state for the caller.
-    set r17 into bonded[r0];
-    // Update the balance of the caller.
-    set r19 into account[r0];
+    // if (new_validator)
+    contains committee[r0] into r17;
+    branch.eq r17 true to validator_in_committee;
+    // {
+        // Check if initial bond amount is at least 100 credits
+        gte r2 100_000_000u64 into r18;
+        // Assert that initial self bond is at least 100 credits
+        assert.eq r18 true;
+
```

Comment:
Why not enforce this above by checking that r10 is greater than 100 credits?

That way we ensure that in the case of existing validators, we guarantee the self-bonded balance is at least 100 credits.

Reply:
Note: We already do this for delegators.

Reply:
I added it b/c to me it made sense to check it in the branch that adds the validator to the committee as it's an initial bond_validator condition. Given the unbond conditions of removing the validator with a self bond of less than 100 credits, it seems like these should both be equivalent.